### PR TITLE
Speed up `GroverOperator` and add GlobalPhase to its decomposition

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
1. The `GroverOperator` is applied by `DefaultQubit` by applying its matrix for up to 12 wires (incl).
However, `GroverOperator` has a lot of structure as it is `2*P-1` for a projector `P`, and a custom application rule is faster than constructing and applying the matrix.
2. The `compute_matrix` method of `GroverOperator` in addition uses a rather contrived way to construct the matrix `np.full((dim, dim), normalization_constant)`. Instead, we can simply use the broadcasting behaviour of numpy and do `normalization_constant - np.eye(dim)` (timings below).
3. The `GroverOperator` decomposition is only correct up to a minus sign (or phase of $\pi$).

**Description of the Change:**
1. Adds a dispatch registration for `apply_operation` in `/qubit/apply_operation.py` for `GroverOperator` that is not based on the matrix representation of `GroverOperator`.
2. Modifies the `compute_matrix` method of `GroverOperator` to skip repetitive calling of `np.kron` and `np.outer` to construct a trivial matrix.
3. Introduces a `GlobalPhase($\pi$)` gate in the decomposition.

**Benefits:**
Speedups and correct decomposition.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

Timing code for the matrix construction:
```python
def f(n_wires):
    s1 = np.array([1, 1]) / np.sqrt(2)
    # uniform superposition state |s>
    s = reduce(np.kron, list(it.repeat(s1, n_wires)))
    # Grover diffusion operator
    return 2 * np.outer(s, s) - np.identity(2**n_wires)

for N in [3, 6, 9, 12]:
    dim = 2**N

    print(f"N={N}")
    %timeit f(N)
    %timeit 2 / dim - np.eye(dim)
    print()
```
image

Timings for the different `apply_operation` options, for three scenarios: Number of wires for operation and state is equal, number of wires for state is that for operation +4, and number of wires for state is double of that for operation.

Given the results, we could do einsum for `len(op.wires)<8` and the dispatch method for bigger wire counts.

Screenshot from 2023-10-12 14-03-56
image
image

### CODE REVIEW

**Review by albi3ro — COMMENTED** [review]
You may need to tweak "use matrix versus decomposition" threshold in default qubit as well.

It is currently located here: https://github.com/PennyLaneAI/pennylane/blob/ea7e1d4e4c5039aec9d4b9c059f1829d7dd10630/pennylane/devices/qubit/preprocess.py#L71

But I have an open PR that will be moving it. We can resolve that merge conflict when we get to it.

**Comment by dwierichs** [comment]
>You may need to tweak "use matrix versus decomposition" threshold in default qubit as well.

Good call. I actually removed it now, because it is supposed to address the large wire counts, which is the regime where the dispatched `apply_grover` is particularly fast. Moreover, as the decomposition also creates a dense matrix (via `MultiControlledX`), this PR makes `GroverOperator` accessible for qubit counts (`>=15` on my laptop) that were inaccessible before :)

> But I have an open PR that will be moving it. We can resolve that merge conflict when we get to it.

:+1:

**Comment by dwierichs** [comment]
Updated timings for all interfaces (line style gives interface, color gives target wires): As there is an outlier in Tensorflow which leads to `num_wires=16`, `num_op_wires=8` having a factor 2 slowdown, still, I will move the threshold from `num_op_wires<8` to `num_op_wires<9`.

image

File: tests/devices/qubit/test_apply_operation.py
Comment (dwierichs) [inline tests/devices/qubit/test_apply_operation.py:873]: Cannot use `math.tensordot` to find out whether `apply_operation_tensordot` was called, because the kernel also uses `math.tensordot`.
`math.argsort` is used by `apply_operation_tensordot` and not by `apply_operation_einsum` or the kernel, so I used that for now.
Is there a proper way to spy on `apply_operation_tensordot` directly?

Comment (albi3ro) [reply tests/devices/qubit/test_apply_operation.py:873]: For testing, I'd mostly just be concerned about "is it correct in all three regimes?".   

For "is it performant in all three regimes?": that's a separate question. We could potentially think about adding a `pytest-benchmark` to this, but it might be overkill at this point in time.

File: tests/templates/test_subroutines/test_grover.py
Comment (albi3ro) [inline tests/templates/test_subroutines/test_grover.py:1]: Mind fixing `test_expand` and the `decomposition_wires` functions?  We should probably also add a check to `test_expand` that the expected decomposition is the same length as the actual decomposition.

Also, note that this test will be getting added soon and will test things like whether or not the matrix matches the decomposition: https://github.com/PennyLaneAI/pennylane/pull/4764/files#diff-73000a2660c03f4c2438c7d18665dcbdd6239b15ec0741137550a712e0d8a98e

File: tests/devices/qubit/test_apply_operation.py
Comment (dwierichs) [reply tests/devices/qubit/test_apply_operation.py:873]: Right :) So is the used placeholder `spy` okay to check whether `apply_operation_tensordot` was called?

File: tests/templates/test_subroutines/test_grover.py
Comment (dwierichs) [reply tests/templates/test_subroutines/test_grover.py:1]: Done :) 
Thanks for the heads up!

File: tests/devices/qubit/test_apply_operation.py
Comment (albi3ro) [inline tests/devices/qubit/test_apply_operation.py:1]: So we have a test checking the output of `apply_operation` when `GroverOperator` has more than nine wires? I see a "it runs without failure" check, but we should also make sure that the answer is the same.

Comment (dwierichs) [reply tests/devices/qubit/test_apply_operation.py:1]: Oh, sorry, I misunderstood your question I think.
I added a check of correctness to the dispatching check (in order to not duplicate computations).

File: tests/devices/qubit/test_apply_operation.py
Comment (timmysilv) [reply tests/devices/qubit/test_apply_operation.py]: I think it's fine. we do something similar to test the `apply_operation` dispatching for ParametrizedEvolution. If the code ever changes, tests will fail and we'll fix 'em!

**Review by timmysilv — COMMENTED** [review]
pretty much good to go, just a few questions

File: doc/releases/changelog-dev.md
Comment (dwierichs) [reply doc/releases/changelog-dev.md:42]: Offline discussions suggested that adding the decomposition could be done in a separate PR. But I'm also fine with holding this one until that is merged, or including it here.

File: tests/devices/qubit/test_apply_operation.py
Comment (dwierichs) [reply tests/devices/qubit/test_apply_operation.py:922]: Nope, does not matter :) I mean, it's not a valid quantum state, but gates are linear maps, so the norm just is left alone.

File: tests/devices/qubit/test_apply_operation.py
Comment (dwierichs) [reply tests/devices/qubit/test_apply_operation.py:928]: Sure, can skip that. I thought it might be nice to have another kernel to check against, but it's not necessary.

File: tests/templates/test_subroutines/test_grover.py
Comment (dwierichs) [inline tests/templates/test_subroutines/test_grover.py:264]:
```suggestion
    assert np.allclose(mat1, mat2)
```

Comment (dwierichs) [reply tests/templates/test_subroutines/test_grover.py:264]: Should i go for the `prod` approach, then? :D

File: tests/devices/qubit/test_apply_operation.py
Comment (timmysilv) [reply tests/devices/qubit/test_apply_operation.py:928]: more just made me wonder if this should be made available to users

Comment (dwierichs) [reply tests/devices/qubit/test_apply_operation.py:925]: Right! I added a small dispatch branch for the full wires case, speeding it up further for that case :)

**Review by timmysilv — APPROVED** [review]
great optimization, thanks for putting this all together! 🎉

File: doc/releases/changelog-dev.md
Comment (trbromley) [reply doc/releases/changelog-dev.md:42]: Yes it'd be great to have the "decompose to nothing" PR done. It's not strictly a blocker to this, but probably best to do it first so we don't forget. @dwierichs is this something you might have time for? (no problem if not, we'll discuss in the core squad).

File: pennylane/devices/qubit/apply_operation.py
Comment (albi3ro) [inline pennylane/devices/qubit/apply_operation.py:300]: If we pulled this out into an `_apply_specialized_grover` function, we would then be able to validate the general process on fewer wires, which would help us a have confidence in at larger system sizes.

Comment (timmysilv) [reply pennylane/devices/qubit/apply_operation.py:300]: agreed, would be nice. we can follow the code standard set by `apply_parametrized_evolution` to be consistent 👍

Comment (dwierichs) [reply pennylane/devices/qubit/apply_operation.py]: Done :)

File: tests/devices/qubit/test_apply_operation.py
Comment (albi3ro) [inline tests/devices/qubit/test_apply_operation.py:863]: Should be no need for copies anymore as `apply_operation` is purely functional.

**Review by albi3ro — APPROVED** [review]

File: tests/devices/qubit/test_apply_operation.py
Comment (dwierichs) [inline tests/devices/qubit/test_apply_operation.py:863]:
```suggestion
        expected_state = self.state
```

**Pull request merged**
